### PR TITLE
Replace the Helm command

### PR DESCRIPTION
**Pull request by nioshield**
### What type of PR is this?
/kind feature

### What this PR does / why we need it:
Use Helm action package directly to manage Helm Releases, avoiding dependency on helm/kustomize Command.The Helm can be flexibly configured
### Which issue(s) this PR fixes:

Fixes #

### Special notes for reviewers:
```
```

### Does this PR introduced a user-facing change?

```release-note
None
```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by wansir** [comment]
@nioshield Thanks for your contribution! The job `pull-kubesphere-verify` is failed, you can get more information about this, refer to https://prow.kubesphere.io/view/s3/prow-logs/pr-logs/pull/kubesphere_kubesphere/4852/pull-kubesphere-verify/1522493840206860288.

File: pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go [hunk 1]
```diff
@@ -470,142 +344,107 @@ func (c *helmWrapper) Install(chartName, chartData, values string) (err error) {
...
-	}
-	defer c.cleanup()
+func (c *helmWrapper) helmUpgrade(chart *chart.Chart, values map[string]interface{}) (*helmrelease.Release, error) {
+	upgrade := action.NewUpgrade(c.helmConf)
+	upgrade.Namespace = c.Namespace
 
-	err = c.setupPostRenderEnvironment()
-	if err != nil {
-		return
+	if c.dryRun || klog.V(8) == true {
```
Comment (xyz-li) [inline pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:361]: I think you should delete `klog.V(8) == true `

File: pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go [hunk 2]
```diff
@@ -470,142 +344,107 @@ func (c *helmWrapper) Install(chartName, chartData, values string) (err error) {
...
+	install := action.NewInstall(c.helmConf)
+	install.ReleaseName = c.ReleaseName
+	install.Namespace = c.Namespace
 
-	// only for mock
-	if c.mock {
-		cmd.Path = os.Args[0]
-		cmd.Args = []string{os.Args[0], "-test.run=TestHelperProcess", "--"}
-		cmd.Env = []string{"GO_WANT_HELPER_PROCESS=1"}
+	if c.dryRun || klog.V(8) == true {
```
Comment (xyz-li) [inline pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:377]: If the user runs ks-apiserver with option `-v 9`, this condition will always be true.

File: pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go [hunk 3]
```diff
@@ -470,142 +344,107 @@ func (c *helmWrapper) Install(chartName, chartData, values string) (err error) {
...
-			"manifest",
-			c.ReleaseName,
-			"--namespace",
-			c.Namespace,
-		},
-		Stderr: stderr,
-		Stdout: stdout,
-	}
+	klog.V(2).Infof("namespace: %s, name: %s, run command success", c.Namespace, c.ReleaseName)
+	klog.V(8).Infof("namespace: %s, name: %s, run command success, release: %v", c.Namespace, c.ReleaseName, rel)
```
Comment (xyz-li) [inline pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:433]: `rel` is a struct, if we use `%v` to format it, the output is difficult to read.

Comment (xyz-li) [reply pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:433]: We should encode `rel` to string.

Comment (nioshield) [reply pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go]: yes,It's enough to output rel.Manifest here

File: pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go [hunk 4]
```diff
@@ -258,60 +192,45 @@ func SetMock(mock bool) Option {
...
+		cfg, _ := clientcmd.RESTConfigFromKubeConfig([]byte(kubeconfig))
+		flags = genericclioptions.NewConfigFlags(false)
+		flags.APIServer = &cfg.Host
+		flags.BearerToken = &cfg.BearerToken
+		flags.CAFile = &cfg.CAFile
+		flags.WithDiscoveryBurst(cfg.Burst)
+		if sa := cfg.Impersonate.UserName; sa != "" {
+			flags.Impersonate = &sa
+		}
+	}
```
Comment (wansir) [inline pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:207]: The following code will cause a cert error if the kubefonfig includes certificate-authority-data.

```
E0513 11:24:17.485511       1 helm_wrapper.go:107] namespace: z1, name: redis-emjr1v, run command failed, error: Kubernetes cluster unreachable: Get "https://10.233.0.1:443/version?timeout=32s": x509: certificate signed by unknown authority
```

Comment (wansir) [inline pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:207]: The following code will cause a cert error if the kubefonfig includes certificate-authority-data.

```
E0513 11:24:17.485511       1 helm_wrapper.go:107] namespace: z1, name: redis-emjr1v, run command failed, error: Kubernetes cluster unreachable: Get "https://10.233.0.1:443/version?timeout=32s": x509: certificate signed by unknown authority
```

Comment (wansir) [reply pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:207]: Wrapped  RESTClientGetter
```
type restClientGetter struct {
	configBytes []byte
}

func (r *restClientGetter) ToRawKubeConfigLoader() clientcmd.ClientConfig {
	config, err := clientcmd.NewClientConfigFromBytes(r.configBytes)
	if err != nil {
		klog.Errorf("failed to create client from bytes: %s", r.configBytes)
	}
	return config
}

func (r *restClientGetter) ToRESTConfig() (*rest.Config, error) {
	return r.ToRawKubeConfigLoader().ClientConfig()
}

func (r *restClientGetter) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
	config, err := r.ToRESTConfig()
	if err != nil {
		return nil, err
	}
	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
	if err != nil {
		return nil, err
	}
	clientset := memory.NewMemCacheClient(discoveryClient)
	return clientset, nil
}
```

Comment (wansir) [reply pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:207]: Wrapped  RESTClientGetter
```
type restClientGetter struct {
	configBytes []byte
}

func (r *restClientGetter) ToRawKubeConfigLoader() clientcmd.ClientConfig {
	config, err := clientcmd.NewClientConfigFromBytes(r.configBytes)
	if err != nil {
		klog.Errorf("failed to create client from bytes: %s", r.configBytes)
	}
	return config
}

func (r *restClientGetter) ToRESTConfig() (*rest.Config, error) {
	return r.ToRawKubeConfigLoader().ClientConfig()
}

func (r *restClientGetter) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
	config, err := r.ToRESTConfig()
	if err != nil {
		return nil, err
	}
	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
	if err != nil {
		return nil, err
	}
	clientset := memory.NewMemCacheClient(discoveryClient)
	return clientset, nil
}
```

Comment (nioshield) [reply pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:207]: Thanks for your reply.in this case,the multi cluster kubeconfig havn't check err,i will fix this err.In addition how can i fast to mock a multi cluster env

Comment (wansir) [reply pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go:207]: > Thanks for your reply.the multi cluster kubeconfig havn't check err,i will fix this err.In addition how can i fast to mock a multi cluster env

`sigs.k8s.io/controller-runtime/pkg/envtest` may be helpful for you.

Comment (nioshield) [reply pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go]: Thanks to your help, I created the MemoryRESTClientGetter to solve this problem

**Comment by wansir** [comment]
LGTM  @xyz-li  Any comments?

**Comment by xyz-li** [comment]
> LGTM @xyz-li Any comments?

Merge after v3.3.0 is released.

**Comment by zheng1** [comment]
Hi, @nioshield have you tried installing the following chart?
https://github.com/kubesphere/helm-charts/tree/master/src/main/mysql
In our previous experience, it may fail because the helm package may not be able to handle the tests folder correctly

**Pull request merged**